docs/Tutorial_DAMP.ipynb

```diff
@@ -0,0 +1,885 @@
+{
```

Comment:
<pre class="ql-syntax" spellcheck="false">Line #11.    print('running time [sec]: ', t_end - t_start)
</pre><p>A few notes regarding running time:</p><p> </p><p>1) IMO, the running time depends on the location of the discord. If it is located to the end of the time series <code>T</code>,then that means the best-so-far value we are looking for will be updated to be the discord distance at the end. Therefore, there is probably a less chance of early-abondoning  in the process of analyzing subsequences from <code>split_idx</code> to that discord index.</p><p> </p><p>2) I think ranomly-generated time series may not be a good choice for comparing the running time between a naive versiona and DAMP version. This is because DAMP looks for discord by considering left neighbors only. As we get closer to the end of time series, a subsequence will have more and more left neighbors. This can increase its chance of having a neighbor to which its distance is less than <code>bsf</code> [and hence not a discord]. That is why we usually see the discord index in the randomly-generated time series is usually close to the <code>split_idx </code>. So, in naive version, it naively computes exact left matrix profile, and in the DAMP, it looks for the discord with early-abondoning. And since th discord index in this case is very close to the beginning, i.e. close to <code>split_idx</code>, we may get higher chance for early-abondoning and hence the running time of DAMP will be very low. [So, a better way might be applying DAMP to a real-world data and see how it performs.]</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/TDAmeritrade/stumpy/pull/872/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='16'/>

Reply:
> I think ranomly-generated time series may not be a good choice for comparing the running time between a naive version and DAMP version

In general, it is always best to report the average/majority case running time rather than the "absolute best case scenario" or the "absolute worst case scenario".

Reply:
Exactly! 

I think randomly-generated time series for discord discovery may give **[almost] absolute best case scenario for DAMP**. Since the top discord is probably located close to the beginning of the time series, the best-so-far discord distance will be updated to the value of true discord distance in the beginning of the process and this results in early abandoning in all of the backward processes of the subsequences that appear later in the time series. I will try to do some experiments.

Reply:
For testing, could we do:

1. Create a sine/cosine wave
2. Add a tiny bit of random noise
3. Insert a discord at any desired location that is composed of random values and see how it changes the computation for different length time series

Reply:
smart! Will do it!

Reply:
@seanlaw 

![image](https://github.com/TDAmeritrade/stumpy/assets/38519522/a5bcc8ba-255c-40c3-98ce-b335c5f7fba8)

```
n = 100001
m = 200 # length of subsequence

seed = 0
np.random.seed(seed)

noise = 0.1 * np.random.rand(n)
T = np.linspace(-500 * np.pi, 500 * np.pi, n)
T[:] = np.sin(T) * (1.0 + noise)

split_idx = 10000
# dummy run
naive_DAMP(T, m, split_idx)
DAMP(T, m, split_idx)

l = n - m + 1
step = 5000
discord_loc = np.arange(split_idx, l, step)

discord = np.random.rand(m)
naive_runnig_time = []
performant_runnig_time = []

for idx in discord_loc:
    T_with_discord = T.copy()
    T_with_discord[idx : idx + m] = discord
    
    ts = time.time()
    naive_DAMP(T_with_discord, m, split_idx)
    te = time.time()
    naive_runnig_time.append(te - ts)
    
    ts = time.time()
    DAMP(T_with_discord, m, split_idx)
    te = time.time()
    performant_runnig_time.append(te - ts)
```

Reply:
The jump in the beginning of the plot of the naive version can be because of njit compilation time. I noticed similar behavior before. So, it is not just that the first call has higher running time because of compilation time. It seems that the higher running time might exist for the first few calls.

The running time of the DAMP increases as the location of the discord gets closer to the end of time series

Reply:
Going to "resolve" this conversation and add the figure to the notebook.